### ISSUE TITLE
Metaclass `__call__`  breaks annotation hover of `__init__` in class using the Metadata

### ISSUE THREAD

**Issue by OlavMSG** [astral-sh/ty#3264]
### Summary

There is a difference in annotation hover and type hint display between ty and Pylance when defining a metaclass with `__call__` and a class defining its own `__init__` and using the metaclass.

Examlpe code:
```py
from typing import Any, TypeVar, reveal_type

T = TypeVar("T")


class TestMeta(type):
    def __call__(cls: type[T], *args: Any, **kwargs: Any) -> T:  # noqa: ANN401
        # expected to have more logic here...
        return super().__call__(*args, **kwargs)  # ty:ignore[invalid-super-argument]


class TestWithMeta(metaclass=TestMeta):
    def __init__(self, val1: str, val2: int, val3: int | None = None) -> None: ...


class TestWithoutMeta:
    def __init__(self, val1: str, val2: int, val3: int | None = None) -> None: ...


value1 = "val1"
value2 = 2
val_test11 = TestWithMeta(value1, value2)
# Displays TestWithMeta(...) on hover in ty
# but class TestWithMeta(
#     val1: str,
#     val2: int,
#     val3: int | None = None
# ) in Pylance
val_test12 = TestWithoutMeta(value1, value2)
# Displays class TestWithoutMeta(
#     val1: str,
#     val2: int,
#     val3: int | None = None
# ) in ty and Pylance
reveal_type(val_test11)  # Runtime type is 'TestWithMeta'
reveal_type(val_test12)  # Runtime type is 'TestWithoutMeta'
```

Schreenshot ty:

<img width="480" height="204" alt="Image" src="https://github.com/user-attachments/assets/9232e03b-843f-4d1c-a27b-3cfa6f12dab5" />

Schreenshot Pylance:

<img width="674" height="266" alt="Image" src="https://github.com/user-attachments/assets/3c642cae-7c2c-4e36-97c0-6b4ec4642acf" />

This Pyright issue seams to be same or similar: https://github.com/microsoft/pyright/issues/5488 

I also assume it may be related to the resent changes because of: https://github.com/astral-sh/ty/issues/2288

### Version

ty 0.0.29

**Comment by sharkdp** [comment]
Thank you for reporting this.

@carljm From a brief look, it seems like we should probably take the signatures of all (up to three) constructors into account when displaying the signature in the LSP, similar to how type context from all constructors is taken into account. Of course, that still leaves the question which signature should be displayed, but it seems reasonable to take the one from the constructor that is called last (here: `__init__`), instead of the one that is being called first?

It feels like ideally, the signatures would be combined somehow (via an intersection), but that seems hard. And it does not seem like that would result in the best type for displaying in this particular example (as it would mix in `Any`).

**Comment by carljm** [comment]
> it seems like we should probably take the signatures of all (up to three) constructors into account

Yes, in some way.

> it seems reasonable to take the one from the constructor that is called last (here: `__init__`), instead of the one that is being called first?

That's certainly the right answer in this particular example, but I feel like maybe it has more to do with the fact that the metaclass `__call__` has the uninformative "gradual signature", or maybe with the fact that it is annotated to return an instance of the class being constructed, rather than a principle of first-called vs last-called?

I feel like this deserves a bit more thorough review of how pylance handles various cases, to see how simple vs complex our heuristics need to be here. It would be really nice if the pylance behavior is actually described by https://typing.python.org/en/latest/spec/constructors.html#converting-a-constructor-to-callable -- though I think that section of the spec is a bit under-specified in terms of what it means to determine (in the abstract, without concrete arguments) whether the return type of a (possibly overloaded and/or generic) constructor method is an "instance of the class being constructed".

**Comment by OlavMSG** [comment]
Just a small follow up.

I am quite new to the use of type checkers. So for all I know my typing in the example above may not be the best for the specific case.

Also thanks for the fast response.
I use uv, ruff and ty in my daily coding.

**Comment by michelesr** [comment]
This makes ty particularly unhelpul when using the AWS CDK library.

With pyright:

<img width="408" height="130" alt="Image" src="https://github.com/user-attachments/assets/a69e006b-0182-4d1b-8898-f0059473ad28" />

With py:

<img width="330" height="132" alt="Image" src="https://github.com/user-attachments/assets/daa76bba-7494-4e49-9acd-da764c0edbd0" />

This is because a lot of CDK classes are using the `JSIIMeta` metaclass.
